packages/backend/src/SynthqlError.ts

```diff
@@ -0,0 +1,139 @@
+import { format } from 'sql-formatter';
+import { DatabaseError } from 'pg';
+import { AnyQuery } from './types';
+
+interface SqlExecutionErrorProps {
+    query: AnyQuery;
+    sql: string;
+    params: readonly any[];
+}
+
+export class SynthqlError extends Error {
+    constructor(
+        public err: Error | any,
+        public type: string,
+        public message: string,
+    ) {
+        super(message);
+
+        Error.captureStackTrace(this, SynthqlError);
+    }
+
+    static createConnectionError({
+        error,
+        url,
+    }: {
+        error: any;
+        url: string;
+    }): SynthqlError {
+        const type = 'ConnectionError';
+
+        const lines = [
+            'Connection error!',
+            '',
+            'Failure to establish a connection to your datastore:',
+            '',
+            url,
+            '',
+            'Check your connection string, and make sure your data store',
+            'is up and can accept new connections',
+        ];
+
+        return new SynthqlError(error, type, lines.join('\n'));
+    }
+
+    static createSqlExecutionError({
+        error,
+        props,
+    }: {
+        error: any;
+        props: SqlExecutionErrorProps;
+    }): SynthqlError {
+        const type = 'SqlExecutionError';
+
+        const message = composeMessage(error, props);
+
+        return new SynthqlError(error, type, message);
+    }
+
+    static createResponseStreamingError({
+        error,
+        query,
+    }: {
+        error: any;
+        query: any;
+    }): SynthqlError {
+        const type = 'ResponseStreamingError';
+
+        const lines = [
+            'Response streaming error!',
+            '',
+            'Failure while streaming back the results for the query:',
+            '',
+            query,
```

Comment:
if query is an object this will serialize to `[Object]`, we should do JSON.strinfigy(query,null,2)